jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1350,8 +1351,22 @@ public void failed(Throwable x)
                 httpChannel = _request.lockedGetHttpChannelState();
                 httpChannel.lockedStreamSendCompleted(false);
             }
-            if (callback != null)
-                httpChannel._writeInvoker.run(() -> HttpChannelState.failed(callback, x));
+            catch (Throwable failure)
+            {
+                httpChannel = null;
+                noChannelFailure = failure;
+            }
+
+            if (httpChannel == null)
+            {
+                // Channel already completed, just fail the callback.
+                HttpChannelState.failed(callback, noChannelFailure);
+            }
+            else
+            {
+                Callback cb = callback;
+                httpChannel._writeInvoker.run(() -> HttpChannelState.failed(cb, x));
```

Comment:
`cb` may be `null`, and `HttpChannelState.failed()` will throw NPE.

Reply:
Not really: if `failed()` is called, this can only mean that `write()` was called and the latter ended up calling `stream.send(..., this)` after storing its `callback` into `_writeCallback`.

But I agree this isn't obvious, so maybe a null check with a warning could be added?

Reply:
`cb` can be null, in case of asynchronous errors.

Calls to `HttpChannelState.onFailure()` for example, call `_response.lockedFailWrite()`, possibly concurrently with a write being finished by NIO.

I think the null check should be restored; if you want to add debug logging great, but it is a normal race that was lost here, and won by an asynchronous failure.

Reply:
Ok, reworked.